File: crates/runtime/src/builtin/webstreams.js
import {
    TextDecoderStream,
    TextEncoderStream,
} from '__wasm_rquickjs_builtin/encoding';

import {
    ByteLengthQueuingStrategy,
    CountQueuingStrategy,
    ReadableByteStreamController,
    ReadableStream,
    ReadableStreamBYOBReader,
    ReadableStreamBYOBRequest,
    ReadableStreamDefaultController,
    ReadableStreamDefaultReader,
    TransformStream,
    TransformStreamDefaultController,
    WritableStream,
    WritableStreamDefaultController,
    WritableStreamDefaultWriter,
} from '__wasm_rquickjs_builtin/streams';

// Patch ReadableStream to throw Node.js-compatible ERR_INVALID_STATE errors
// when the stream is already locked.
function errInvalidState(msg) {
    const e = new TypeError(msg);
    e.code = 'ERR_INVALID_STATE';
    return e;
}

function errInvalidArgValue(msg) {
    const e = new RangeError(msg);
    e.code = 'ERR_INVALID_ARG_VALUE';
    return e;
}

function errArgNotIterable() {
    const e = new TypeError('The "iterable" argument must be an instance of Iterable.');
    e.code = 'ERR_ARG_NOT_ITERABLE';
    return e;
}

const origGetReader = ReadableStream.prototype.getReader;
ReadableStream.prototype.getReader = function getReader(...args) {
    try {
        return origGetReader.apply(this, args);
    } catch (err) {
        if (err instanceof TypeError && err.code === undefined && /locked/i.test(err.message)) {
            throw errInvalidState('Invalid state: ReadableStream is locked');
        }
        throw err;
    }
};

const origValues = ReadableStream.prototype.values;
if (typeof origValues === 'function') {
    ReadableStream.prototype.values = function values(...args) {
        try {
            return origValues.apply(this, args);
        } catch (err) {
            if (err instanceof TypeError && err.code === undefined && /locked/i.test(err.message)) {
                throw errInvalidState('Invalid state: ReadableStream is locked');
            }
            throw err;
        }
    };
}

const symAsyncIterator = Symbol.asyncIterator;
const origAsyncIterator = ReadableStream.prototype[symAsyncIterator];
if (typeof origAsyncIterator === 'function') {
    ReadableStream.prototype[symAsyncIterator] = function(...args) {
        try {
            return origAsyncIterator.apply(this, args);
        } catch (err) {
            if (err instanceof TypeError && err.code === undefined && /locked/i.test(err.message)) {
                throw errInvalidState('Invalid state: ReadableStream is locked');
            }
            throw err;
        }
    };
}

const origByobReaderRead = ReadableStreamBYOBReader.prototype.read;
ReadableStreamBYOBReader.prototype.read = function read(view, ...args) {
    return Promise.resolve(origByobReaderRead.call(this, view, ...args)).catch((err) => {
        if (err instanceof TypeError && err.code === undefined && /zero-length|non-zero|detached/i.test(err.message)) {
            throw errInvalidState('Invalid state: View or Viewed ArrayBuffer is zero-length or detached');
        }
        throw err;
    });
};

const origRespondWithNewView = ReadableStreamBYOBRequest.prototype.respondWithNewView;
ReadableStreamBYOBRequest.prototype.respondWithNewView = function respondWithNewView(view) {
    try {
        return origRespondWithNewView.call(this, view);
    } catch (err) {
        if (err instanceof RangeError && err.code === undefined) {
            throw errInvalidArgValue(err.message);
        }
        if (err instanceof TypeError && err.code === undefined && /detached|invalidated|state|closed|zero-length|non-zero/i.test(err.message)) {
            throw errInvalidState(err.message);
        }
        throw err;
    }
};

ReadableStream.from = function from(iterable) {
    if (iterable == null) {
        throw errArgNotIterable();
    }
    const asyncIteratorFactory = iterable[Symbol.asyncIterator];
    const iteratorFactory = iterable[Symbol.iterator];
    if (typeof asyncIteratorFactory !== 'function' && typeof iteratorFactory !== 'function') {
        throw errArgNotIterable();
    }

    let iterator;
    return new ReadableStream({
        async pull(controller) {
            if (iterator === undefined) {
                iterator = typeof asyncIteratorFactory === 'function'
                    ? asyncIteratorFactory.call(iterable)
                    : iteratorFactory.call(iterable);
            }
            const result = await iterator.next();
            if (result.done) {
                controller.close();
            } else {
                controller.enqueue(result.value);
            }
        },
        async cancel(reason) {
            if (iterator && typeof iterator.return === 'function') {
                await iterator.return(reason);
            }
        },
    });
};

export {
    ByteLengthQueuingStrategy,
    CountQueuingStrategy,
    ReadableByteStreamController,
    ReadableStream,
    ReadableStreamBYOBReader,
    ReadableStreamBYOBRequest,
    ReadableStreamDefaultController,
    ReadableStreamDefaultReader,
    TextDecoderStream,
    TextEncoderStream,
    TransformStream,
    TransformStreamDefaultController,
    WritableStream,
    WritableStreamDefaultController,
    WritableStreamDefaultWriter,
};

File: crates/runtime/src/builtin/webstreams.rs
// JS functions for the web streams implementation
pub const WEBSTREAMS_JS: &str = include_str!("web-streams-polyfill-4.1.0.js");

// JS wrapper that patches ReadableStream for Node.js compatibility
pub const WEBSTREAMS_WRAPPER_JS: &str = include_str!("webstreams.js");

// Re-export for aliases
pub const REEXPORT_JS: &str = r#"export * from '__wasm_rquickjs_builtin/webstreams_wrapper';
import * as _all from '__wasm_rquickjs_builtin/webstreams_wrapper';
export default _all;"#;

// JS code wiring the web streams module into the global context
pub const WIRE_JS: &str = r#"
        import {
            ByteLengthQueuingStrategy as __ByteLengthQueuingStrategy,
            CountQueuingStrategy as __CountQueuingStrategy,
            ReadableByteStreamController as __ReadableByteStreamController,
            ReadableStream as __ReadableStream,
            ReadableStreamBYOBReader as __ReadableStreamBYOBReader,
            ReadableStreamBYOBRequest as __ReadableStreamBYOBRequest,
            ReadableStreamDefaultController as __ReadableStreamDefaultController,
            ReadableStreamDefaultReader as __ReadableStreamDefaultReader,
            TransformStream as __TransformStream,
            TransformStreamDefaultController as __TransformStreamDefaultController,
            WritableStream as __WritableStream,
            WritableStreamDefaultController as __WritableStreamDefaultController,
            WritableStreamDefaultWriter as __WritableStreamDefaultWriter,
        } from '__wasm_rquickjs_builtin/webstreams_wrapper';
        globalThis.ByteLengthQueuingStrategy = __ByteLengthQueuingStrategy;
        globalThis.CountQueuingStrategy = __CountQueuingStrategy;
        globalThis.ReadableByteStreamController = __ReadableByteStreamController;
        globalThis.ReadableStream = __ReadableStream;
        globalThis.ReadableStreamBYOBReader = __ReadableStreamBYOBReader;
        globalThis.ReadableStreamBYOBRequest = __ReadableStreamBYOBRequest;
        globalThis.ReadableStreamDefaultController = __ReadableStreamDefaultController;
        globalThis.ReadableStreamDefaultReader = __ReadableStreamDefaultReader;
        globalThis.TransformStream = __TransformStream;
        globalThis.TransformStreamDefaultController = __TransformStreamDefaultController;
        globalThis.WritableStream = __WritableStream;
        globalThis.WritableStreamDefaultController = __WritableStreamDefaultController;
        globalThis.WritableStreamDefaultWriter = __WritableStreamDefaultWriter;
    "#;
